File: day3/src/main/java/PartTwo.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PartTwo {
    public static boolean active = true;

    public static int main(char[][] input) {
        return Arrays.stream(input)
                        .map(line -> extract(line))
                        .mapToInt(line -> multiply(line))
                        .sum();
    }

    public static ArrayList<String> extract(char[] line) {

        ArrayList<String> cleaned = new ArrayList<>();

        Pattern pattern = Pattern.compile("mul\\(\\d{1,3},\\d{1,3}\\)|do\\(\\)|don\\'t\\(\\)");
        Matcher matcher = pattern.matcher(new String(line));
        
        while(matcher.find()) {
            cleaned.add(matcher.group());
        }
        System.out.println(cleaned);

        return cleaned;
    }

    public static int multiply (ArrayList<String> strings) {
        int result = 0;
        for (String s: strings) {
            if (s.equals("don't()")) {
                active = false;
                continue;
            }

            if (s.equals("do()")) {
                active = true;
            }

            if (active && !s.equals("do()") && !s.equals("don't()")) {
                // split it on the ','
                String[] halves = s.split(",");
                // remove everything that isn't a number
                String left = halves[0].replaceAll("[^0-9]", "");
                String right = halves[1].replaceAll("[^0-9]", "");
                // multiply each number
                result += (Integer.parseInt(left) * Integer.parseInt(right));
    
            }
       }

        return result;
    }
}

File: day10/part_two.py
with open('day10.input', 'r') as file:
    grid = [[int(char)  if char != '.' else float('inf') for char in line.split()[0]] for line in file]

result = 0

def dfs(r, c, prev):
    if (r > -1 and c > -1 and r < len(grid) and c < len(grid[0]) and grid[r][c] - prev == 1):
        if grid[r][c] == 9:
            return 1
        return sum([
            dfs(r + 1, c, grid[r][c]), dfs(r, c + 1, grid[r][c]),
            dfs(r - 1, c, grid[r][c]), dfs(r, c - 1, grid[r][c])
        ])
    else:
        return 0 
    
for r in range(len(grid)):
    for c in range(len(grid[0])):
        if grid[r][c] == 0:
            result += dfs(r, c, grid[r][c] - 1)

print(result)